### PR TITLE
Tests: Revamp static bwc test framework to use dangling indexes

### PR DESCRIPTION
**Pull request by rjernst**
The static old index tests currently take a long time to run because
each index version essentially recreates the cluster, and spins up
new nodes.  This PR instead loads each old version into the existing
cluster as a dangling index. It also removes the intermediate
"StaticIndexBackwardCompatibilityTest" which was an extra layer
with no purpose, and moves a shared version of a commonly found
function to get an http client.

The test now takes between 40 and 60 seconds for me. I also ran it
"under stress" by running all ES tests in one shell, while
simultaneously running 10 iterations of the old index tests. Each
iteration took on average about 90 seconds, which is much better
than the 20+ minutes we see in master on jenkins.

### CODE REVIEW

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 1]
```diff
@@ -19,46 +19,60 @@
...
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
 
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 
-@TimeoutSuite(millis = 40 * TimeUnits.MINUTE)
-public class OldIndexBackwardsCompatibilityTests extends StaticIndexBackwardCompatibilityTest {
+@LuceneTestCase.SuppressCodecs({"Lucene3x", "MockFixedIntBlock", "MockVariableIntBlock", "MockSep", "MockRandom", "Lucene40", "Lucene41", "Appending", "Lucene42", "Lucene45", "Lucene46", "Lucene49"})
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: out of curiosity  -  why are all these suppresses needed?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 2]
```diff
@@ -70,6 +84,76 @@ public static void initIndexes() throws Exception {
...
+    public Settings nodeSettings(int ord) {
+        return ImmutableSettings.builder()
+            .put(Node.HTTP_ENABLED, true) // for _upgrade
+            .put(MergePolicyModule.MERGE_POLICY_TYPE_KEY, NoMergePolicyProvider.class) // disable merging so no segments will be upgraded
+            .build();
+    }
+    
+    void setupDataLoadingNodes() throws Exception {
+        Path dataDir = newTempDirPath(LifecycleScope.SUITE);
+        indicesDir = dataDir.resolve(internalCluster().getClusterName()).resolve("nodes/0/indices/");
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: why not let the test cluster manage this and do something ala `internalCluster().getInstance(NodeEnvironment.class, node_1).nodeDataPaths()`;  to retrieve the chosen path?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 3]
```diff
@@ -70,6 +84,76 @@ public static void initIndexes() throws Exception {
...
+        ImmutableSettings.Builder nodeSettings = ImmutableSettings.builder()
+            .put("path.data", dataDir.toAbsolutePath());
+        internalCluster().startNode(nodeSettings);
+    }
+
+    String loadIndex(String indexFile) throws Exception {
+        Path unzipDir = newTempDirPath();
+        Path unzipDataDir = unzipDir.resolve("data");
+        String indexDir = indexFile.replace(".zip", "");
+        String indexName = indexDir.toLowerCase(Locale.ROOT);
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: wondering - why is this needed?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 4]
```diff
@@ -70,6 +84,76 @@ public static void initIndexes() throws Exception {
...
+        assertTrue(Files.exists(unzipDataDir));
+        Path[] list = FileSystemUtils.files(unzipDataDir);
+        if (list.length != 1) {
+            throw new IllegalStateException("Backwards index must contain exactly one cluster");
+        }
+        
+        // move to the real data dir
+        Path src = list[0].resolve("nodes/0/indices/" + indexName);
+        Path dest = indicesDir.resolve(indexName);
+        /*StringBuffer msg = new StringBuffer();
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: left over?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 5]
```diff
@@ -70,6 +84,76 @@ public static void initIndexes() throws Exception {
...
+        }
+        assertTrue(msg.toString(), Files.exists(src));
+        */assertTrue("[" + indexFile + "] missing index dir: " + src.toString(), Files.exists(src));
+        Files.move(src, dest);
+        assertFalse(Files.exists(src));
+        assertTrue(Files.exists(dest));
+        
+        // force reloading dangling indices with a cluster state republish
+        client().admin().cluster().prepareReroute().get();
+        ensureGreen(indexName);
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I think there is a race condition here - it may take some time for the dangling logic to kick in (it's async via network calls). I would assertBusy until the index is visible in the cluster state, then go into ensure green.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 6]
```diff
@@ -70,6 +84,76 @@ public static void initIndexes() throws Exception {
...
+    
+    @Override
+    public Settings nodeSettings(int ord) {
+        return ImmutableSettings.builder()
+            .put(Node.HTTP_ENABLED, true) // for _upgrade
+            .put(MergePolicyModule.MERGE_POLICY_TYPE_KEY, NoMergePolicyProvider.class) // disable merging so no segments will be upgraded
+            .build();
+    }
+    
+    void setupDataLoadingNodes() throws Exception {
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: this starts just one node, right? can we do setupDataLoadingNode (no S)?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 7]
```diff
@@ -97,25 +181,25 @@ public void testAllVersionsTested() throws Exception {
...
 
     public void testOldIndexes() throws Exception {
+        setupDataLoadingNodes();
+        
         Collections.shuffle(indexes, getRandom());
         for (String index : indexes) {
+            long startTime = System.currentTimeMillis();
             logger.info("Testing old index " + index);
             assertOldIndexWorks(index);
+            logger.info("Done with " + index + ", took " + ((System.currentTimeMillis() - startTime)/1000.0) + " seconds");
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: can we log with a "-->" prefix? i.e.,  "--> done ... " . Makes it easier to find & grep.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 8]
```diff
@@ -139,54 +233,42 @@ void assertBasicSearchWorks() {
         ElasticsearchAssertions.assertNoFailures(searchReq.get());
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: this confuses me - do you internationally search twice?

Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: never mind, I see the sort now. Can we add a logger.info("--> testing basic search"), logger.info("--> testing basic search with sort")?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 9]
```diff
@@ -139,54 +233,42 @@ void assertBasicSearchWorks() {
...
-        client().prepareUpdate("test", "doc", docId).setDoc("foo", "bar").get();
-        GetResponse getRsp = client().prepareGet("test", "doc", docId).get();
+        client().prepareUpdate(indexName, "doc", docId).setDoc("foo", "bar").get();
+        GetResponse getRsp = client().prepareGet(indexName, "doc", docId).get();
         Map<String, Object> source = getRsp.getSourceAsMap();
         assertThat(source, Matchers.hasKey("foo"));
 
-        assertAcked(client().admin().indices().prepareUpdateSettings("test").setSettings(ImmutableSettings.builder()
+        assertAcked(client().admin().indices().prepareUpdateSettings(indexName).setSettings(ImmutableSettings.builder()
             .put("refresh_interval", "1s")
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: can we use org.elasticsearch.index.engine.EngineConfig#DEFAULT_REFRESH_INTERVAL & org.elasticsearch.index.shard.IndexShard#INDEX_REFRESH_INTERVAL ?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 10]
```diff
@@ -139,54 +233,42 @@ void assertBasicSearchWorks() {
...
             .build()));
-        SearchRequestBuilder searchReq = client().prepareSearch("test").setQuery(QueryBuilders.matchAllQuery());
+        SearchRequestBuilder searchReq = client().prepareSearch(indexName).setQuery(QueryBuilders.matchAllQuery());
         SearchHit hit = searchReq.get().getHits().getAt(0);
         String docId = hit.getId();
         // foo is new, it is not a field in the generated index
-        client().prepareUpdate("test", "doc", docId).setDoc("foo", "bar").get();
-        GetResponse getRsp = client().prepareGet("test", "doc", docId).get();
+        client().prepareUpdate(indexName, "doc", docId).setDoc("foo", "bar").get();
+        GetResponse getRsp = client().prepareGet(indexName, "doc", docId).get();
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: can we add a test that turns off real time and checks you get the old doc back?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 11]
```diff
@@ -139,54 +233,42 @@ void assertBasicSearchWorks() {
...
-            .put("number_of_replicas", 0))
-            .execute().actionGet());
-        waitNoPendingTasksOnAll(); // make sure the replicas are removed before going on
+        logger.debug("Creating [{}] replicas for index [{}]", numReplicas, indexName);
+        assertAcked(client().admin().indices().prepareUpdateSettings(indexName).setSettings(ImmutableSettings.builder()
+            .put("number_of_replicas", numReplicas)
+        ).execute().actionGet());
+        ensureGreen(indexName);
+        
+        // TODO: do something with the replicas! query? index?
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: +1 :)

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 6 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I now better understand how it works. I don't think we need to start a node at all. Just pick a random node from the 3 the test infra setups for you and capture it's data path. I wonder if we even need this method in that case. We can probably make do with loadIndex alone , no?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java [hunk 12]
```diff
@@ -139,54 +233,42 @@ void assertBasicSearchWorks() {
...
+    void assertUpgradeWorks(String indexName, boolean alreadyLatest) throws Exception {
         HttpRequestBuilder httpClient = httpClient();
 
         if (alreadyLatest == false) {
-            UpgradeTest.assertNotUpgraded(httpClient, "test");
+            UpgradeTest.assertNotUpgraded(httpClient, indexName);
         }
-        UpgradeTest.runUpgrade(httpClient, "test", "wait_for_completion", "true");
-        UpgradeTest.assertUpgraded(httpClient, "test");
+        UpgradeTest.runUpgrade(httpClient, indexName);
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: wondering - what happend to wait_for_completion ?

**Comment by bleskes** [comment]
This is much cleaner! thx . Left some comments.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 1 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I don't remember, but they were copied from the static index superclass.  I believe @s1monw added them originally on the first static bwc test, so maybe he can comment.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 2 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: That just seems more complicated than is necessary.  The only reason I have other nodes here is for testing replication.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 3 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Index names cannot contain upper case letters, but the zip file names are based on the elasticsearch version, which can contain `Beta` and `RC`.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 4 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Yes, removed.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 5 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I just tested this out, and it looks like the cluster health api is "smart" about this. I ran a cluster health on a non existent index, and the request blocks, eventually timing out.  I then ran the request again, but this time creating the index while the request was ongoing, and the request returned, not timed out.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 12 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: wait_for_completion was removed from the upgrade api altogether in #9640

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 8 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Added

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 7 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Sure, added.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 9 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I changed the value to use the constant, since that could change, but I think using the actual parameter inline makes it much easier to read through test code than using constants everywhere.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 10 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I would rather leave adding more checks for #9143 and keep this issue focusing on improving the test performance.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 6 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I named this method plural with future enhancements in mind. My plan is to actually start 2 nodes, one with a single data path, and another with 2 data paths.  loadIndex can then randomly choose to put the index on the single data path node, or split the index files across the 2 data paths node.  This will help test the upgrade path for the behavior change to remove the DistributorDirectory.

**Comment by rjernst** [comment]
@bleskes I pushed another commit with some changes based on your feedback.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 5 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Yeah, it seems it is. We treat non existing indices as red. Thx for educating me.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 10 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Up to you. it seems to me that if we do test real time get, we should do it properly. If you choose not to do it, can you add a note on #9143 so it won't be lost?

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 2 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: @rjernst this is what I meant : https://github.com/bleskes/elasticsearch/commit/1e38fac385757fac62f0ab0532d557d804caddab 

Imho it's cleaner as it full relies on the test cluster to setup things. If we randomize to mutilpe data paths (which we should, it will randomly select one for the injection. Note that the commit also have formatting fixed, so it will be bigger).

**Comment by bleskes** [comment]
Thx @rjernst . I replied to the comments.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 2 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: btw - the test uncovered some issue with the dangling indices import. You might run into a node not connected issues

Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Regarding the randomization, I'm trying very hard *not* make this test highly randomized.  I don't think the purpose of this test should be to test random configurations, etc.  I would like debugging to be very clear and not surprising.  Having a single node path (or 2 sets of paths in the future) is much more direct and debuggable (I will add a logging line for the path(s)).

Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Sure, then just pick the first node and the first path.

Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Then I would have to make some choice in nodeSettings for which node will have multiple data paths in the future.

Is there some technical problem with what I have? We are talking about a difference of less than 10 lines of code, and it seems to be strictly a matter of personal preference.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 10 above)
Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: I've added a note on #9143.

File: src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java (hunk 2 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Well, obviously this is not about the line of codes.  

It better integrates with the internal cluster + the NodeEnvironment choice's of data paths. No hard coded reference to it or the node ordinal. It creates node data where it is supposed to be, as opposed to a random place somewhere etc (I can come up with more if need be :) ) . 

If in the future we want to introduce any randomness (which imho is a good thing here because you want to see that bwc works under different node environments, but that's another discussion), we practically get it for free.

Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: > It creates node data where it is supposed to be, as opposed to a random place somewhere

This is not correct? Each node created by the cluster uses a random data dir. It is no different, except that I keep a hard reference to it, instead of looking it up after the internal test cluster creates the random dir.

Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: > If in the future we want to introduce any randomness (which imho is a good thing here because you want to see that bwc works under different node environments, but that's another discussion), we practically get it for free.

IMO we use too much randomness: randomness is hard to debug, and having direct tests for things would catch 97% of the errors (67% of stats are made up!). I only mean that we have tests for checking randomness at the cluster level.  This test is supposed to check upgrading an old index, and I want to keep it focused on that goal.  Adding lots of randomness may find other, unrelated errors, but I would rather leave that to other tests whose sole purpose is to spit out lots of randomness to find those errors.

Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: > This is not correct? Each node created by the cluster uses a random data dir.

All the cluster nodes put their data next to each other, under the current working dir. After you run the tests, check for example: `target/J0/data/` .  As I said, I really don't think you should remap data paths here. Just use what the test cluster decided for you. Not starting an extra node is a bonus.

Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Sorry, I got confused because I have a branch I've been working on which fixes this writing to CWD. We shouldn't have any code relying on CWD, and with my change, each node will have its own isolated base path.

> Not starting an extra node is a bonus.

There is no extra node? The data nodes started as part of the cluster are just for replicas. 

Is this disagreement on test style really worth blocking this improvement? Perhaps in the future I will agree, but we should get the tests faster now, and work on perfection later.

Comment (s1monw) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: i have to agree with @rjernst there is no point of adding the unnecessary variant of a random node that is already running here. I think it's important for tests like this to reduce the variance. these tests should be to the point. we are not testing dangling indices here.

Comment (bleskes) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Let me re-iterated what my concerns are regarding my current approach
- It overrides default path handling of the InternalTestCluster without needing to.
- It overrides path logic in NodeEnvironment w.r.t where to put the data. NodeEnvironment expose the API to get it.
- It starts another node where we can make it simpler and use the async standard node start logic of InternalTestCluster (minor)

My point here was not w.r.t randomness but rather making the code simpler and straight forward.

Comment (rjernst) [inline src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityTests.java]: Regarding paths: if we don't explicitly set path.data, the test node setup will randomly choose to make this multiple data paths. I am not ready to add multiple data path support here yet.

As I said before, I am not necessarily against the possibility of using a precreated node/path in the future, but right now I think it adds more complication we shouldn't have to deal with yet.

**Comment by rjernst** [comment]
@bleskes I pushed a new commit.  I believe this addresses your concern over using hard coded paths.

**Comment by bleskes** [comment]
Awesome. LGTM. I will work on the dangling request issue today, so we can get this in.

**Comment by bleskes** [comment]
scratch the waiting for the dangling indices request issue. missed the master:false on the loading node. No need to wait indeed.
